msos/WaitChains.cs

```diff
@@ -17,68 +18,561 @@ class WaitChains : ICommand
+        {
+            List<UnifiedBlockingObject> criticalSectionObjects = new List<UnifiedBlockingObject>();
+
+            foreach (var item in stack)
+            {
+                UnifiedBlockingObject blockObject;
+
+                if (_stackWalker != null && _stackWalker.GetCriticalSectionBlockingObject(item, out blockObject))
                 {
-                    if (owner == null) // ClrMD sometimes reports this nonsense
-                        continue;
+                    criticalSectionObjects.Add(blockObject);
+                }
+            }
+
+            if (criticalSectionObjects.Any())
+            {
+                if (list == null)
+                    list = new List<UnifiedBlockingObject>();
+
+                list.AddRange(criticalSectionObjects);
+            }
+        }
+
+        internal virtual List<UnifiedBlockingObject> GetUnmanagedBlockingObjects(
+            List<UnifiedStackFrame> unmanagedStack, uint engineThread)
+        {
+            unmanagedStack.ForEach(stackFrame => _stackWalker?.SetFrameParameters(stackFrame, _runtime));
+
+            List<UnifiedBlockingObject> result = new List<UnifiedBlockingObject>();
+
+            var framesWithHandles = unmanagedStack.Where(frame => frame.Handles?.Count > 0);
+
+            foreach (var frame in framesWithHandles)
+            {
+                foreach (var handle in frame.Handles)
+                {
+                    result.Add(new UnifiedBlockingObject(handle.Value, handle.ObjectName, handle.Type));
+                }
+            }
+
+            CheckForCriticalSections(result, unmanagedStack);
+
+            return result;
+        }
+
+        internal List<UnifiedStackFrame> ConvertToUnified(IEnumerable<DEBUG_STACK_FRAME> stackFrames,
+            ClrRuntime runtime, ThreadInfo info)
+        {
+            var reversed = stackFrames.Reverse();
+            List<UnifiedStackFrame> stackTrace = new List<UnifiedStackFrame>();
 
-                    DisplayChainForThreadAux(owner, depth + 2, visitedThreadIds);
+            foreach (var frame in reversed)
+            {
+                var unified_frame = new UnifiedStackFrame(frame, (IDebugSymbols2)_debugClient);
+
+                _stackWalker?.SetFrameParameters(unified_frame, runtime);
+
+                stackTrace.Add(unified_frame);
+            }
+
+            return stackTrace;
+        }
+    }
+
+
+    /// <summary>
+    /// LiveProcessStrategy UnifiedBlockingObjects fetching:
+    ///     native threads (x86): MiniDump, StackWalker
+    ///     native threads (x64): MiniDump
+    ///     managedthreads (x86/x64): ClrMd
+    /// </summary>
+    class DumpFileBlockingObjectsStrategy : BlockingObjectsStrategy
+    {
+        public DumpFileBlockingObjectsStrategy(CommandExecutionContext context) : base(context)
+        {
+            _miniDump = new DumpReader(context.DumpFile);
```

Comment:
What do you mean by "isn't updated"? You need to add `EnumerateHandles` (or whatever) to it so that you can call it from msos. For the other implementations of `IDataReader`, throw a `NotImplementedException` for now -- we can think later if it's worth implementing or not.